### DISCUSSION TITLE
16. May 2023

### DISCUSSION THREAD

**Discussion by dimakuv** [gramineproject/gramine#1339] [Meeting notes]
### Agenda

*(please write your proposed agenda items in comments under this discussion)*

- [x] Dmitrii: new API in RA-TLS to extract additional data (in addition to `/dev/attestation/quote`)? Case in point: the JSON Web Token (JWT) sent back by Microsoft Azure Attestation, see https://github.com/gramineproject/contrib/pull/38
    - Proposal 1: simply use envvar like `RA_TLS_MAA_JWT` (no need for changes in core Gramine repo)
    - Proposal 2: extend `ra_tls_verify_callback_results` with a pointer to such blobs (needs changes in core Gramine repo)
    - Proposal 3: new RA-TLS API (function) to extract such blobs (needs changes in core Gramine repo)
    - (*this is a continuation of discussion started on [25. April 2023](https://github.com/gramineproject/gramine/discussions/1292)*)

- [x] Dmitrii: removing support for Ubuntu 18.04 completely
   - We already removed 18.04 [from our Jenkins CI](https://github.com/gramineproject/gramine/commit/a9c48a59fe4d82cb9a966b12080b08b1a2c4715f)
   - Do we want to proactively remove support in Gramine proper? This would include:
      - subprojects (mbedTLS patch)
      - documentation
      - Gramine codebase
      - Tests/examples
      - GSC
      - ...anything else?

- [x] Dmitrii: Syscall entry point in Gramine and XSAVE preservation problem
   - See https://github.com/gramineproject/gramine/issues/1352
   - See https://github.com/gramineproject/gramine/blob/4946a0b2224eb989faec3c2e81d8cb3458a9e853/libos/src/arch/x86_64/syscallas.S#L14-L19

---

#### Benny: documentation reorganization PRs

Benny: [1 huge PR](https://github.com/gramineproject/gramine/pull/1241) was split into several smaller PRs: 2 PRs [already](https://github.com/gramineproject/gramine/pull/1346) [merged](https://github.com/gramineproject/gramine/pull/1347), 1 PR [pending](https://github.com/gramineproject/gramine/pull/1348), then 1 more PR.

Woju: side note but the "linear commit history" decision is not working that well now. Woju votes for allowing merge commits. This is offtopic anyway, this discussion could be postponed for later.

#### Dmitrii: new API in RA-TLS to extract additional data

[ Dmitrii describes the problem as described in #1292 ]

Mona: this assumes that Gramine already got the JWT (token) from some peer/verifier. But how do we get the JWT from the peer in the first place?

Woju: What Dmitrii describes is that the token is the authentication credential that is used after the attestation, to gain access to some other resources.

Woju: Are there similar approaches to gain access to resources? Dmitrii: the only thing I know about is [Edgeless Marblerun Premain](https://pkg.go.dev/github.com/edgelesssys/marblerun/marble/premain) trick; but this is only at startup and not generic enough.

Woju: I'd like to put things that are not part of the hardware contract/spec **not** under `/dev/attestation/`. So Woju would keep it outside of core Gramine.

Woju: we can have a switch for RA-TLS generation of X.509 cert: that it shouldn't hard-code SGX quote path `/dev/attestation/quote` but it can have any path. So this path doesn't need to be under `/dev/attestation/`, but it can be e.g. a tmpfs file `/tmp/gramine/token`, populated beforehand by the app itself or by the RA-TLS variant.

Mona: with Interoperable RA-TLS, we have an extensible format of the evidence inside the X.509 cert. So we could ask the Interoperable RA-TLS committee to add new evidence types like MAA JWT tokens. Then RA-TLS would need to also get the evidence type as the input (maybe a new function argument? or something like this).

Woju: we'd like to keep the RA-TLS general code (that is developed together with core Gramine) to be generic and usable by all those different plugins and verification schemes. So looks like we don't need changes in core Gramine, but need some generic changes to RA-TLS.

Dmitrii: Please check the Interoperable RA-TLS standard: https://confidentialcomputing.io/2023/03/06/unifying-remote-attestation-protocol-implementations/.

#### Removing support for Ubuntu 18.04 completely

Mariusz and Scott: can be aggressive and remove all the quirks of 18.04.

**Comment by mkow** [comment]
// I couldn't join the call so I only know what's in these notes.

> Woju: side note but the "linear commit history" decision is not working that well now.

How is that related to this case? I don't see any connection. In that case someone just put 10 unrelated changes inside a single commit, it doesn't matter if we merged or rebased it.

**Reply by dimakuv** [reply]
I think it wasn't related at all. But something in our conversation triggered Woju to make his claim on the inadequacy of linear commit history. Maybe @woju remembers and could explain better.

**Reply by woju** [reply]
This was in response to general comment that the docs team has trouble cleaning up the reorganisation PR. There are three problems this PR have, one is sheer size, which is just what it is, second is splitting into commits (which needs to be done anyway, because that's good practice and allows to separately review separate concerns like moving vs editing), and the last one is this one, that it needs linear history. Maintaining linear history with the simultaneous frowning upon frequent rebasing makes it yet more problematic to maintain the branch especially if it's very big.

I've also said this is to be discussed after the docs reorganisation is done, not before.

**Comment by monavij** [comment]
On fully dropping Ubuntu 18.04 support, let's announce our plan on gitter channel to reach active users and see if anyone complains.

**Reply by dimakuv** [reply]
@monavij I think we were told to hold off with the "aggressive" plan for now?

**Reply by monavij** [reply]
> @monavij I think we were told to hold off with the "aggressive" plan for now?

Yes you are right. We did agree to wait until release 1.6 to fully drop support and also start giving a warning that it will be deprecated.

**Reply by mkow** [reply]
> We did agree to wait until release 1.6 to fully drop support

We did agree to stop maintaining 18.04 starting from 1.5. When and who agreed to postpone this? :/

**Reply by dimakuv** [reply]
> > We did agree to wait until release 1.6 to fully drop support
> 
> We did agree to stop maintaining 18.04 starting from 1.5. When and who agreed to postpone this? :/

During the other meeting :) It's not in the Tuesday meeting notes...

**Reply by woju** [reply]
I think I said during this (16.05.2023) meeting there will be no 1.5 packages for 18.04.

**Reply by woju** [reply]
I also think that Scott said SDK and/or other core SGX projects drop 18.04, because we don't care about "extended support". And I don't recall that anyone objected to merging patches that would break 18.04, even because of lack of testing.

Anyway, ISTM we probably need to discuss this yet again in tomorrow (23.05.2023) call.

**Reply by kailun-qin** [reply]
> I think I said during this (16.05.2023) meeting there will be no 1.5 packages for 18.04.

I think we discussed this long ago and have achieved consensus on no package for 18.04 (?). But what's not clear is around "proactively removing support for 18.04"/"stop maintaining 18.04 starting from 1.5"?

**Reply by dimakuv** [reply]
Yes, as @kailun-qin explained.

**Reply by mkow** [reply]
For me "stop maintaining 18.04 starting from 1.5" sounds right, but how do you define "aggressively dropping it" exactly? What if we merge something which accidentally breaks it?

Also, the meeting notes seem to indicate that you actually agreed to fully drop 18.04 (Scott and Mariusz in favor, no notes about any objections; I'm also in favor, I guess woju is also ;) ).

**Reply by dimakuv** [reply]
> For me "stop maintaining 18.04 starting from 1.5" sounds right, but how do you define "aggressively dropping it" exactly? What if we merge something which accidentally breaks it?

I think we won't have anything like this. IIUC, Alpine changes could break this, but I haven't seen anything that would truly break Ubuntu 18.04?

> Also, the meeting notes seem to indicate that you actually agreed to fully drop 18.04 (Scott and Mariusz in favor, no notes about any objections; I'm also in favor, I guess woju is also ;) ).

Yes, this happened on Tuesday meeting (where these meeting notes are from). But on another Wednesday meeting, we had some additional feedback that 18.04 should be tried to be kept in v1.5.

**Reply by woju** [reply]
> I haven't seen anything that would truly break Ubuntu 18.04?

https://reviewable.io/reviews/gramineproject/gramine/1197#-NPI6meE9RlxjzPh27si

Last time I've checked this thread, we were okay with merging a thing that will traceback `gramine-sgx-sign` in the face of the user on 18.04.

**Reply by dimakuv** [reply]
@woju But this PR won't cut it into v1.5? At least we didn't slate it for v1.5 (which should happen quite soon).

**Reply by mkow** [reply]
I think it will, it's almost ready and I see no reason to artificially postpone it (_especially not because of 18.04_).

Btw. Why do we even care about such users? If they don't care about updating their distro, then why would they care about updating Gramine? They can use the old, unsupported Gramine releases same as they use old, unsupported Ubuntu. And actually nothing will actively break for them, `apt update` just won't pull Gramine 1.5 and stay on v1.4.

**Reply by dimakuv** [reply]
This conversation is at a dead-end :) Let's discuss it again during the Tuesday meeting, with @monavij giving her view on the Ubuntu 18.04 support decision...
